### ISSUE TITLE
Reproducible jars with a LICENSE.txt file

### ISSUE THREAD

**Issue by carlosame** [mozilla/rhino#1120]
Currently, every time that Rhino is built, different jar files are produced even for the same set of code. Which could be an annoyance if the packages were posted to additional Maven repositories (for example a wokflow that posted to Github packages), because the digests of those files would be different.

I'd suggest configuring the Gradle build for reproducible jar files, which would imply removing the `Built-Date`and `Built-Time` fields from `MANIFEST.MF`.

I'd also copy the `LICENSE.txt` file into the `META-INF` directory of each jar file (the license is currently being copied to the root directory of the main `rhino-<version>.jar` file, and not the others).

**Comment by p-bakker** [comment]
Wouldn't each built jar by definition be different, as we built when the source code changes?

**Comment by carlosame** [comment]
> Wouldn't each built jar by definition be different, as we built when the source code changes?

The idea is that if I build the source code at a specific git tag (or release), it is good that the produced jar files are unique. Currently they aren't.

I mean if today, two people at different places built packages for Rhino 1.7.14-RC1, they would be different and with different MD5/SHA*/etc digests.

More information on the motivations for reproducible builds:

- https://reproducible-builds.org/
- https://en.wikipedia.org/wiki/Reproducible_builds

**Comment by p-bakker** [comment]
Thnx for those links, make sense to me

**Comment by carlosame** [comment]
I could provide a PR with the necessary changes to `build.gradle`, although that's only part of the process: a reproducible build requires a JDK that's reasonably close to the one used to publish the artifacts.

For example, it is known that the `ZipOutputStream`/`JarOutputStream` in Oracle JDKs produce Zip/Jar files which are slightly different to their OpenJDK counterparts, even for very close Java versions.

**Comment by p-bakker** [comment]
> although that's only part of the process: a reproducible build requires a JDK that's reasonably close to the one used to publish the artifacts.

That's not something we can solve though, no?

What about publishing checksums of the builds we provide? Would that be useful?

**Comment by carlosame** [comment]
> That's not something we can solve though, no?

No, Rhino would only do "its part" in providing a build that can be reproduced (albeit not always reproducible with any toolchain). The people performing the reproducible build should care about the rest (assuming e.g. that a recent JDK 8/11/17 was used).

> What about publishing checksums of the builds we provide? Would that be useful?

The Gradle build is already doing that, look at the Maven Central distibution of 1.7.14-RC1:

https://repo1.maven.org/maven2/org/mozilla/rhino/1.7.14-RC1/

There are a number of `*.jar.md5`, `*.jar.sha1`, `*.jar.sha256` and `*.jar.sha512` which are digests of the various jar files; neither the jar files nor the digests are currently reproducible.

Incidentally, the digest verification model of Maven repositories is "interesting": the digests come from the same server, even the same directory where the checksummed files are... so if an attacker managed to alter the binaries they'd presumably also modify the digests... But that's a different issue.

What could be done by Rhino is to mention which Java version was used to deploy to Maven Central. And if Rhino is ever published to (currently useless) Github packages (I could provide a PR for that automated release task as well), try to use the same JDK version that was used to deploy to Maven Central.

**Comment by tonygermano** [comment]
If it was really important to people, couldn't they verify the files *in* the jar? The .class files should all be reproducible already, no (assuming the same build tools are used?)

**Comment by carlosame** [comment]
> If it was really important to people, couldn't they verify the files _in_ the jar? The .class files should all be reproducible already, no (assuming the same build tools are used?)

That would imply publishing digests for _all_ the .class files in the Jar, and verifying all of them. Not very practical.

My concerns are related to the digests that are already being produced, and the changes required to fix those digests in the build file are really modest.

**Comment by p-bakker** [comment]
> The Gradle build is already doing that, look at the Maven Central distibution of 1.7.14-RC1

Sry, hadn't looked closely enough

>  And if Rhino is ever published to (currently useless) Github packages

I had been looking at the Github Packages feature, but couldn't really wrap my head around it :-) When you say useless, you mean in general or because we're not using it right now?

>  could provide a PR with the necessary changes to build.gradle

For my part I'd welcome such PR, I only see an upside, no downside, but not sure if others have different opinions. 

> I could provide a PR for that automated release task as well

Heck, why not :-)

While at it: feel like tackling these ones as well? #1119, #957 & #956 :-) all in the same area somewhat

**Comment by carlosame** [comment]
> I had been looking at the Github Packages feature, but couldn't really wrap my head around it :-) When you say useless, you mean in general or because we're not using it right now?

I mean in general, as mentioned in https://github.com/mozilla/rhino/issues/980#issuecomment-881408831, because Github Packages requires authentication to download the packages.

And the last time that I looked at it, the package owner needed to authorize individually each user (downloader). That's utterly impractical and quite a few people at Github think the same, just look at the github.community link that I provided in the aforementioned comment.

> For my part I'd welcome such PR, I only see an upside, no downside, but not sure if others have different opinions.

I'll do.

> > I could provide a PR for that automated release task as well
> 
> Heck, why not :-)
> 
> While at it: feel like tackling these ones as well? #1119 & #956 :-) all in the same area somewhat

I was unaware of #956 and is clearly part of what I wanted to do here. I could add a commit to deal with #1119 as well, although that one is very new and can still be subject to a lot of bikeshedding... I'll leave that commit for the last minute :-D

**Comment by tonygermano** [comment]
I *think* with github packages you can create a read-only auth token, that can then be shared publicly with the instructions for downloading from the repo.

**Comment by carlosame** [comment]
> I _think_ with github packages you can create a read-only auth token, that can then be shared publicly with the instructions for downloading from the repo.

Perhaps one can do that, but my downstream users do not even want to hear about authenticated access to packages.

**Comment by tonygermano** [comment]
It doesn't seem like too much of an inconvenience to me that if I'm needing to set up a custom repo anyway to also copy/paste the auth info.

**Comment by p-bakker** [comment]
piggybacking on this issue, as it's related to the build setup: there's this file in the repo: https://github.com/mozilla/rhino/blob/master/man/rhino.1

It got added with commit https://github.com/mozilla/rhino/commit/ef999cfd8b62783487b4f382f9bd4f4ffc132d2a#diff-48b676e2b107da679512b793d5fd4cc4329f0c7c17a97cf6e0e3d1005b600b03, to resolve this bug https://bugzilla.mozilla.org/attachment.cgi?id=744851&action=diff

The only reference I can find to this file is here, in build.gradle: https://github.com/mozilla/rhino/blob/master/build.gradle#L459

However, I don't see the file ending up in any output of the build process anywhere and not sure if the reference in build.gradle  is just something obsolete or whether it should be doing something.

Any thoughts?

**Comment by carlosame** [comment]
> However, I don't see the file ending up in any output of the build process anywhere and not sure if the reference in build.gradle is just something obsolete or whether it should be doing something.

That reference is not working but I could fix it. What needs to be decided is whether the project wants to ship a Rhino manpage or not. (although this discussion really belongs to #1119)

**Comment by carlosame** [comment]
After the recent merge of reproducible build commits, now only the modification of `.gitattributes` remains as a pending task (to close this issue completely). That modification should probably be done at the same time as the global `spotlessApply` that is being planned.

**Comment by p-bakker** [comment]
@carlosame Are there still changes to be made to the .gitattributes? I see changes were made to it 16 days ago

**Comment by carlosame** [comment]
> @carlosame Are there still changes to be made to the .gitattributes? I see changes were made to it 16 days ago

Yes the .gitattributes still needs to be changed. As mentioned in my previous post, that should be done in the same PR that applies the planned big 'spotlessApply'.

Feel free to copy the .gitattributes and .editorconfig in https://github.com/mozilla/rhino/pull/1124#discussion_r769095217, if anyone is willing to send that PR.
